runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/common/worker/MapTaskExecutor.java

```diff
@@ -146,13 +165,17 @@ public boolean supportsRestart() {
   @Override
   public void abort() {
     // Signal the read loop to abort on the next record.
-    // TODO: Also interrupt the execution thread.
     for (Operation op : operations) {
       Preconditions.checkState(op instanceof ReadOperation || op instanceof ReceivingOperation);
       if (op instanceof ReadOperation) {
         ((ReadOperation) op).abortReadLoop();
       }
     }
+    synchronized (this) {
+      if (currentExecutorThread != null) {
+        currentExecutorThread.interrupt();
```

Comment:
There is no requirement for user code to handle being interrupted arbitrarily and there are enough instances that I have seen where this is not handled gracefully by the user.

Reply:
Can you give an example to consider here?

Reply:
My understanding of this change is that user code will typically finish a callback and the abort will occur when control returns to the worker. To a user it should be transparently like other failures. Is that not the case?

Reply:
Interrupting the GCS writer has caused issues with how the pipe that is used to transfer data is not gracefully shutdown leaving a blocked thread sitting around indefinitely.

Reply:
Internally within Flume we saw with the C++ worker where user code didn't handle thread::cancel correctly and were relying on process crash to not get stuck.

I would generally love for us to be able to interrupt random code arbitrarily as it would be the best and cleanest way to do this.

Reply:
I may be misunderstanding, but this is a cooperative interrupt. The interrupt only happens when the user code return control.

Reply:
Discussed offline: "user" code can mean many things, and there are some common Java errors that this change could trigger:

1. User's (who usually won't check the bit) doing catch-all error handling without separating out InterruptedException.
2. IO libraries (which may often check the bit) doing incomplete or incorrect cleanup and being left in an inconsistent state or leaking resources.

These would ideally both be noticed and handled by higher-level mechanisms since in both cases something should be notably unhealthy about the thread, process, or VM. For now just aborting without interrupting the thread is safer.